### PR TITLE
fix(txpool/blob): make blob delete failed more accurate

### PR DESCRIPTION
**Pull request by jsvisa**
From the metric of `reth_transaction_pool_blobstore_failed_deletes` I saw too many blob failed:

image

Then run the reth node with `-vvvv` debug mode, found all the blob delete failed logs was caused by file not found:

```bash
2024-07-26T12:38:06.848876Z DEBUG txpool::blob: err=[0xc4dedce22285ec1747b2a916d32c2106c59bf55701d6f18193bec639519b6f6a] failed to delete blob file at /data/blobstore/c4dedce22285ec1747b2a916d32c2106c59bf55701d6f18193bec639519b6f6a: No such file or directory (os error 2)
2024-07-26T12:38:06.848888Z DEBUG txpool::blob: err=[0xf8acf98e52632ff89cc76e07d3095bfb7be27ea3aed6a2432e30a8fc7f8070d8] failed to delete blob file at /data/blobstore/f8acf98e52632ff89cc76e07d3095bfb7be27ea3aed6a2432e30a8fc7f8070d8: No such file or directory (os error 2)
2024-07-26T12:38:06.848900Z DEBUG txpool::blob: err=[0xf1cb35953efd2868ed049f6fd1d43b675e0f18daef5be6f979eac4cc9efcba7e] failed to delete blob file at /data/blobstore/f1cb35953efd2868ed049f6fd1d43b675e0f18daef5be6f979eac4cc9efcba7e: No such file or directory (os error 2)
```

After some digging, I think maybe the blob sidecar was inserted in the txpool module:
https://github.com/paradigmxyz/reth/blob/1ffa3d147df51328e27140faf7038b090c0b4955/crates/transaction-pool/src/pool/mod.rs#L429-L458

however, some blobs were not propagated into our node, so the blob wasn't tracked in the blobstore.

But the to be cleaned the blobs were extracted from the blocks, the results were mucher than our tracked ones.

https://github.com/paradigmxyz/reth/blob/1ffa3d147df51328e27140faf7038b090c0b4955/crates/transaction-pool/src/blobstore/tracker.rs#L36-L67

### CODE REVIEW

File: crates/transaction-pool/src/blobstore/disk.rs [hunk 1]
```diff
@@ -62,12 +62,16 @@ impl BlobStore for DiskFileBlobStore {
...
+            self.inner.txs_to_delete.write().insert(tx);
+        }
         Ok(())
     }
 
     fn delete_all(&self, txs: Vec<B256>) -> Result<(), BlobStoreError> {
-        self.inner.txs_to_delete.write().extend(txs);
+        for tx in txs {
+            let _ = self.delete(tx);
+        }
```
Comment (mattsse) [inline crates/transaction-pool/src/blobstore/disk.rs:74]: we can optimize this slightly by doing a specialized impl here,
we only want to acquire the locks once

so we should add another function to inner that is

```rust
fn retain_existing(&self, v: &mut Vec) {
  // acquire lock once then do v.retain
  // then retain if exists on disk
}
```

https://github.com/paradigmxyz/reth/blob/cf7698ab035bcb8209ea6cd422feace7ce37fc47/crates/transaction-pool/src/blobstore/disk.rs#L226-L232

File: crates/transaction-pool/src/maintain.rs [hunk 2]
```diff
@@ -188,20 +188,19 @@ pub async fn maintain_transaction_pool<Client, P, St, Tasks>(
...
+            {
+                metrics.inc_deleted_tracked_blobs(blobs.len());
+                // remove all finalized blobs from the blob store
+                pool.delete_blobs(blobs);
+                // and also do periodic cleanup
+                let pool = pool.clone();
+                task_spawner.spawn_blocking(Box::pin(async move {
+                    debug!(target: "txpool", finalized_block = %finalized, "cleaning up blob store");
+                    pool.cleanup_blobs();
+                }));
```
Comment (mattsse) [inline crates/transaction-pool/src/maintain.rs:202]: style: could you please undo this change here, I like this match to be exhaustive here because this highlights that all cases are covered

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, ty.

one suggestion and style nit

File: crates/transaction-pool/src/maintain.rs (hunk 2 above)
Comment (jsvisa) [reply crates/transaction-pool/src/maintain.rs:202]: But if no blobs were found in the block range, I think no need to spawn a task to cleanup, correct me if I'm wrong or anything I'm missing.

Comment (mattsse) [reply crates/transaction-pool/src/maintain.rs:202]: good point, I missed that change, then we can move L199-L204 into the finalized arm

File: crates/transaction-pool/src/blobstore/disk.rs [hunk 3]
```diff
@@ -231,6 +234,16 @@ impl DiskFileBlobStoreInner {
...
     }
 
+    /// Returns all the blob transactions which are in the cache or on the disk.
+    fn retain_existing(&self, txs: Vec<B256>) -> Result<Vec<B256>, BlobStoreError> {
+        let mut cache = self.blob_cache.lock();
+        Ok(txs
+            .iter()
+            .filter(|&tx| cache.get(tx).is_some() || self.blob_disk_file(*tx).is_file())
+            .cloned()
+            .collect())
```
Comment (mattsse) [inline crates/transaction-pool/src/blobstore/disk.rs:244]: hmm, this is a bit more tricky then I thought.

what we should actually do here is 
a first over over the cache with

https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.partition

then drop the cache lock
and then a second pass over those that weren't in the cache via `self.blob_disk_file(*tx).is_file()`

because we don't wont to hold the cache lock while we do IO

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/transaction-pool/src/blobstore/disk.rs (hunk 3 above)
Comment (jsvisa) [reply crates/transaction-pool/src/blobstore/disk.rs]: @mattsse ptal, thanks

**Review by mattsse — APPROVED** [review]
